src/common/telemetry/src/tracing_sampler.rs

```diff
@@ -0,0 +1,205 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use opentelemetry::trace::{
+    Link, SamplingDecision, SamplingResult, SpanKind, TraceContextExt, TraceId, TraceState,
+};
+use opentelemetry::KeyValue;
+use opentelemetry_sdk::trace::{Sampler, ShouldSample};
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(default)]
+pub struct TracingSampleOptions {
+    pub default_ratio: f64,
+    pub rules: Vec<TracingSampleRule>,
+}
+
+impl Default for TracingSampleOptions {
+    fn default() -> Self {
+        Self {
+            default_ratio: 1.0,
+            rules: vec![],
+        }
+    }
+}
+
+#[derive(Clone, Default, Debug, Serialize, Deserialize)]
+#[serde(default)]
+pub struct TracingSampleRule {
+    pub protocol: String,
+    pub request_types: Vec<String>,
+    pub ratio: f64,
+}
+
+impl PartialEq for TracingSampleOptions {
+    fn eq(&self, other: &Self) -> bool {
+        self.default_ratio == other.default_ratio && self.rules == other.rules
+    }
+}
+impl PartialEq for TracingSampleRule {
+    fn eq(&self, other: &Self) -> bool {
+        self.protocol == other.protocol
+            && self.request_types == other.request_types
+            && self.ratio == other.ratio
+    }
+}
+
+impl Eq for TracingSampleOptions {}
+impl Eq for TracingSampleRule {}
+
+pub fn create_sampler(opt: &TracingSampleOptions) -> Box<dyn ShouldSample> {
+    if opt.rules.is_empty() {
+        Box::new(Sampler::TraceIdRatioBased(opt.default_ratio))
+    } else {
+        Box::new(RuleSampler::new(opt))
+    }
+}
+
+/// Determine the sampling rate of a span according to the `rules` provided in `RuleSampler`.
+/// For spans that do not hit any `rules`, the `default_ratio` is used.
+#[derive(Debug, Clone)]
+struct RuleSampler {
+    default_ratio: f64,
+    rules: Vec<Rule>,
+}
+
+#[derive(Debug, Clone)]
+struct Rule {
+    protocol: String,
+    request_types: HashSet<String>,
+    ratio: f64,
+}
+
+impl Rule {
+    pub fn new(rule: &TracingSampleRule) -> Self {
+        Self {
+            protocol: rule.protocol.clone(),
+            request_types: rule.request_types.iter().cloned().collect(),
+            ratio: rule.ratio,
+        }
+    }
+
+    pub fn match_rule(&self, protocol: &str, request_type: Option<&str>) -> Option<f64> {
+        if protocol == self.protocol {
+            if self.request_types.is_empty() {
+                Some(self.ratio)
+            } else if request_type.is_none() {
+                None
+            } else if self.request_types.contains(request_type.unwrap()) {
+                Some(self.ratio)
+            } else {
+                None
+            }
```

Comment:
Can the codes rewrite like this? It seems more understandable:
```suggestion
            if self.request_types.is_empty() {
                Some(self.ratio)
            } else {
                if let Some(t) = request_type && self.request_types.contains(t) {
                    Some(self.ratio)
                } else {
                    None
                }
            }
```